qiskit/transpiler/passes/synthesis/unitary_synthesis.py

```diff
@@ -137,99 +142,200 @@ def run(self, dag: DAGCircuit) -> DAGCircuit:
             Output dag with UnitaryGates synthesized to target basis.
 
         Raises:
-            TranspilerError:
-                1. pulse_optimize is True but pulse optimal decomposition is not known
-                   for requested basis.
-                2. pulse_optimize is True and natural_direction is True but a preferred
-                   gate direction can't be determined from the coupling map or the
-                   relative gate lengths.
+            QiskitError: if a 'method' was specified for the class and is not
+                found in the installed plugins list. The list of installed
+                plugins can be queried with
+                :func:`~qiskit.transpiler.passes.synthesis.plugins.unitary_synthesis_plugin_names`
         """
+        if not self.method:
+            method = "default"
+        else:
+            method = self.method
+        if method not in self.plugins.ext_plugins:
+            raise QiskitError("Specified method: %s not found in plugin list" % method)
+        plugin_method = self.plugins.ext_plugins[method].obj
+        if plugin_method.supports_coupling_map:
+            dag_bit_indices = {bit: idx for idx, bit in enumerate(dag.qubits)}
+        for node in dag.named_nodes(*self._synth_gates):
+            synth_dag = None
+            kwargs = {}
+            if plugin_method.supports_basis_gates:
+                kwargs["basis_gates"] = self._basis_gates
+            if plugin_method.supports_coupling_map:
+                kwargs["coupling_map"] = self._coupling_map
+                kwargs["qubits"] = [dag_bit_indices[x] for x in node.qargs]
+            if plugin_method.supports_approximation_degree:
+                kwargs["approximation_degree"] = self._approximation_degree
+            if plugin_method.supports_natural_direction:
+                kwargs["natural_direction"] = self._natural_direction
+            if plugin_method.supports_pulse_optimize:
+                kwargs["pulse_optimize"] = self._pulse_optimize
+            if plugin_method.supports_gate_lengths:
+                gate_lengths = {}
+                if self._backend_props:
+                    for gate in self._backend_props._gates:
+                        gate_lengths[gate] = {}
+                        for k, v in self._backend_props._gates[gate].items():
+                            length = v.get("gate_length")
+                            if length:
+                                gate_lengths[gate][k] = length[0]
+                        if not gate_lengths[gate]:
+                            del gate_lengths[gate]
+                kwargs["gate_lengths"] = gate_lengths
```

Comment:
8 levels of indentation might be a record! Is it worth pulling out this logic into a separate function, just to flatten it out a little?